README.md
# nlp-engine

The NLP engine is the engine that runs text analyzers writtein in [NLP++](http://visualtext.org). It can be called in two different ways:

1. Calling the nlp.exe command line executable (this is what the VSCode NLP++ Language Extension does)
1. Calling from within C++ or another language that can call c++ functions

# Command Line

You can get help on nlp.exe:

      [command arg: --help]

      usage: nlp [--version] [--help]
                 [-INTERP][-COMPILED] INTERP is the default
                 [-ANA analyzer] name or path to NLP++ analyzer folder
                 [-IN infile] input text file path
                 [-OUT outdir] output directory
                 [-WORK workdir] working directory
                 [-DEV][-SILENT] /DEV generates logs, -SILENT (default) does not
                 [infile [outfile]] when no /IN or -OUT specified

      Directories in the nlp.exe files:
         data        nlp engine bootstrap grammar
         analyzers   default location for nlp++ analyzer folders
         visualtext  common files for the VisualText IDE

## Switches

Switch | Function
------------ | -------------
-INTERP / -COMPILED | Runs NLP++ code interpreted or compiled
-ANA | name of the analyzer or path to the analyzer folder
-IN | Input file
-OUT | Output directory


qdbm/ChangeLog
2006-11-02  Mikio Hirabayashi  <mikio@users.sourceforge.net>

	- A bug related to optimization on Intel Mac series was escaped.
	- Release: 1.8.74

2006-10-20  Mikio Hirabayashi  <mikio@users.sourceforge.net>

	- A bug related to BZIP2 encoding was fixed.
	- Release: 1.8.73

2006-10-06  Mikio Hirabayashi  <mikio@users.sourceforge.net>

	- The utility API was enhanced.
	- Release: 1.8.72

2006-08-24  Mikio Hirabayashi  <mikio@users.sourceforge.net>

	- Makefile was modified to assure portability.
	- Silent flag was added to the C++ API, the Java API, and the Ruby API.
	- Release: 1.8.71

2006-08-18  Mikio Hirabayashi  <mikio@users.sourceforge.net>

	- A bug of segmentation fault on BSD was fixed.
	- A test command for checking multi thread safety was added.
	- Release: 1.8.70

2006-08-15  Mikio Hirabayashi  <mikio@users.sourceforge.net>

	- Mutex controll in C++ API became refined.
	- Release: 1.8.69

2006-08-08  Mikio Hirabayashi  <mikio@users.sourceforge.net>

	- A bug about memory alignment was fixed.
	- A bug of handling meta data on big endian platforms was fixed.
	- Release: 1.8.68

2006-08-05  Mikio Hirabayashi  <mikio@users.sourceforge.net>

	- A bug about memory alignment was fixed.
	- A bug about parsing MIME was fixed.
	- Release: 1.8.67

2006-08-05  Mikio Hirabayashi  <mikio@users.sourceforge.net>

	- The utility API was enhanced.
	- Release: 1.8.66

2006-08-03  Mikio Hirabayashi  <mikio@users.sourceforge.net>

	- The extended API was enhanced.
	- The extended advanced API was enhanced.
	- Release: 1.8.65

2006-07-28  Mikio Hirabayashi  <mikio@users.sourceforge.net>

	- A bug of Makefile ralated to optimization was fixed.
	- Release: 1.8.64

2006-07-24  Mikio Hirabayashi  <mikio@users.sourceforge.net>

	- A lot of functions were replaced by macros.
	- The utility API was enhanced.
	- Release: 1.8.63

2006-07-20  Mikio Hirabayashi  <mikio@users.sourceforge.net>

	- A lot of functions were replaced by macros.
	- The URL resolver was to allow unescaped meta characters.
	- The advanced API was enhanced.
	- Release: 1.8.62

2006-07-14  Mikio Hirabayashi  <mikio@users.sourceforge.net>

	- A bug of the size checking function of the extended API was fixed.
	- The advanced API was enhanced.
	- Release: 1.8.61

2006-06-03  Mikio Hirabayashi  <mikio@users.sourceforge.net>

	- The basic API was enhanced.
	- The extended API was enhanced.
	- The advanced API was enhanced.
	- Multiple cursor class is now supported on the java API.
	- Release: 1.8.60

2006-05-30  Mikio Hirabayashi  <mikio@users.sourceforge.net>

	- The basic API was enhanced.
	- The extended API was enhanced.
	- The advanced API was enhanced.
	- A bug of Makefile for Mac OS X support was fixed.
	- Release: 1.8.59

2006-05-28  Mikio Hirabayashi  <mikio@users.sourceforge.net>

	- The basic API was enhanced.
	- The advanced API was enhanced.
	- Release: 1.8.58

2006-05-20  Mikio Hirabayashi  <mikio@users.sourceforge.net>

	- The basic API was enhanced.
	- The utility API was enhanced.
	- Release: 1.8.57

2006-05-17  Mikio Hirabayashi  <mikio@users.sourceforge.net>

	- A bug of URL decoder was fixed.
	- Release: 1.8.56

2006-05-15  Mikio Hirabayashi  <mikio@users.sourceforge.net>

	- The extended API was enhanced.
	- The utility API was enhanced.
	- Release: 1.8.55

2006-05-15  Mikio Hirabayashi  <mikio@users.sourceforge.net>

	- The basic API was enhanced.
	- Release: 1.8.54

2006-05-10  Mikio Hirabayashi  <mikio@users.sourceforge.net>

	- AIX is now supported.
	- The utility API was enhanced.
	- Release: 1.8.53

2006-05-04  Mikio Hirabayashi  <mikio@users.sourceforge.net>

	- A bug about evaluating RFC822 date format was fixed.
	- Release: 1.8.52

2006-05-02  Mikio Hirabayashi  <mikio@users.sourceforge.net>

	- A bug about evaluating RFC822 date format was fixed.
	- Warings from GCC 4.1.x were dealt with.
	- Missing functions in qdbm.def were supplied.
	- Release: 1.8.51

2006-04-28  Mikio Hirabayashi  <mikio@users.sourceforge.net>

	- The utility API was enhanced.
	- Release: 1.8.50

2006-04-19  Mikio Hirabayashi  <mikio@users.sourceforge.net>

	- A parameter of dynamic linking library on Mac OS X was modified.
	- The utility API was enhanced.
	- Release: 1.8.49

2006-03-27  Mikio Hirabayashi  <mikio@users.sourceforge.net>

	- The utility API was enhanced.
	- Release: 1.8.48

2006-03-10  Mikio Hirabayashi  <mikio@users.sourceforge.net>

	- LTmakefile was modified.
	- The utility API was enhanced.
	- Release: 1.8.47

2006-02-20  Mikio Hirabayashi  <mikio@users.sourceforge.net>

	- The utility API was enhanced.
	- Release: 1.8.46

2006-01-28  Mikio Hirabayashi  <mikio@users.sourceforge.net>

	- Alignment algorithm was improved.
	- A bug of mmap emulation on Windows was fixed.
	- Release: 1.8.45

2006-01-24  Mikio Hirabayashi  <mikio@users.sourceforge.net>

	- A bug of handling meta data on big endian platforms was fixed.
	- The advanced API was enhanced.
	- Release: 1.8.44

2006-01-24  Mikio Hirabayashi  <mikio@users.sourceforge.net>

	- A bug of mmap emulation on Windows was fixed.
	- Release: 1.8.43

2006-01-22  Mikio Hirabayashi  <mikio@users.sourceforge.net>

	- mmap emulation on Windows was enhanced.
	- Release: 1.8.42

2006-01-13  Mikio Hirabayashi  <mikio@users.sourceforge.net>

	- Compression of pages of B+ tree with LZO and BZIP was added.
	- Release: 1.8.41

2006-01-08  Mikio Hirabayashi  <mikio@users.sourceforge.net>

	- Configuration for VC++ was to build DLL.
	- Release: 1.8.40

2006-01-04  Mikio Hirabayashi  <mikio@users.sourceforge.net>

	- The advanced API was enhanced.
	- Release: 1.8.39

2006-01-01  Mikio Hirabayashi  <mikio@users.sourceforge.net>

	- The advanced API was enhanced.
	- Release: 1.8.38

2005-12-26  Mikio Hirabayashi  <mikio@users.sourceforge.net>

	- Functions to set the size of the free block pool were added.
	- Commands line tools were enhanced.
	- Release: 1.8.37

2005-12-24  Mikio Hirabayashi  <mikio@users.sourceforge.net>

	- Free block pool was enhanced.

lite/str.h
/*******************************************************************************
Copyright (c) 1998-2010 by Text Analysis International, Inc.
All rights reserved.
********************************************************************************
*
* NAME:	STR.H
* FILE:	lite\str.h
* CR:		11/04/98 AM.
* SUBJ:	String and char functions.
*
*******************************************************************************/

#ifndef STR_H_
#define STR_H_

class Parse;	// Forward reference.

#define SPELL_BUF_LEN 10001

LITE_API bool alphabetic(_TCHAR ch);										// 09/22/99 AM.
LITE_API bool is_upper(_TCHAR ch);											// 12/16/01 AM.
LITE_API bool is_lower(_TCHAR ch);											// 12/16/01 AM.
LITE_API _TCHAR to_upper(_TCHAR ch);											// 12/16/01 AM.
LITE_API _TCHAR to_lower(_TCHAR ch);											// 12/16/01 AM.

bool str_to_int(_TCHAR *, /*UP*/ int &);
bool str_to_long(_TCHAR *, /*UP*/ long &);
inline int char_to_digit(_TCHAR ch);
bool all_uppercase(_TCHAR *str);
bool all_lowercase(_TCHAR *str);										// 01/03/00 AM.
bool mixcap(_TCHAR *str);													// 01/15/00 AM.

_TCHAR *str_to_lower(_TCHAR *str, _TCHAR *buf);
_TCHAR *str_to_lower(_TCHAR *str);                   // 12/15/14 AM.
_TCHAR *str_to_upper(_TCHAR *str, _TCHAR *buf);							// 11/04/99 AM.
_TCHAR *str_to_upper(_TCHAR *str);                   // 12/15/14 AM.
bool zap_final_white(_TCHAR *str);										// 02/12/99 AM.

bool str_equal(_TCHAR *str1, _TCHAR *str2);							// 06/07/99 AM.

bool long_to_str(long num, /*UP*/ _TCHAR *buf);					// 11/03/99 AM.

bool str_to_cstr(															// 12/15/99 AM.
	_TCHAR *str,			// String to convert.
	/*UP*/
	_TCHAR *buf			// Supplied buffer to fill.
	);
inline void strcpy_e(														// 12/30/99 AM.
	_TCHAR* &ptr,			// Pointer to buffer for copying into.
							// Points to last FILLED char.
	_TCHAR *str			// Terminated string to copy.
	);
//inline void strcpy_ee(														// 09/27/00 AM.
//	char* &ptr,			// First EMPTY spot.
//	char *str
//	);
bool clean_str(
	_TCHAR *str,			// String to clean.
	/*UP*/
	_TCHAR *buf			// Supplied buffer to fill.
	);

bool find_str(_TCHAR *, const _TCHAR **);									// 05/17/00 AM.
bool find_str_nocase(_TCHAR *, const _TCHAR **);							// 06/07/00 AM.

bool strlen_eq(_TCHAR *, long);												// 05/24/00 AM.
bool strlen_eq(long, _TCHAR *);												// 05/24/00 AM.
bool strlen_eq(_TCHAR *, _TCHAR *);											// 05/24/00 AM.
long binary_search(															// 06/09/00 AM.
	_TCHAR *str,				// String to search.
	_TCHAR *arr[],			// Array to search.
	long len					// Active length of array (excluding 0th elt).
	);
long binary_search(															// 06/09/00 AM.
	_TCHAR *str,				// String to search.
	_TCHAR *arr[],			// Array to search.
	long ii,					// "Left" end of array.
	long jj					// "Right" end of array.
	);

// FUNCTIONS TO CONVERT FILES WITH JUNK IN THEM.					// 07/18/00 AM.
// Trying much longer lines by default.                              // 10/07/05 AM.
//bool readConvert(_TCHAR *,_TCHAR* &,long &,long=MAXSTR);				// 07/18/00 AM.
//bool readConvertU(_TCHAR *,_TCHAR* &,long &,long=MAXSTR);          // 03/03/05 AM.
bool readConvert(_TCHAR *,_TCHAR* &,long &,long=MAXLINE);				// 10/07/05 AM.
bool readConvertU(_TCHAR *,_TCHAR* &,long &,long=MAXLINE);           // 10/07/05 AM.
bool UTL_GetUnicodeFromMBCS(                                         // 03/03/05 AM.
	wchar_t* &pszUnicodeString,
	LPCTSTR lpszMBCS,
	long &len,
	UINT nCodePage
	);


bool writeConvert(_TCHAR *,_TCHAR *,long=0,long=MAXSTR);				// 07/18/00 AM.

bool UTL_GetMBCSFromUnicodeString(                                   // 03/06/05 AM.
	LPCWSTR lpwUnicode,
	UINT nCodePage,
	LPCTSTR* &pszMultiByteString
	);


_TCHAR *str_to_title(_TCHAR *str, _TCHAR *buf);								// 08/21/00 DD.
_TCHAR *str_wrap(_TCHAR *str, long wrapSize, _TCHAR *buf);           // 09/22/00 DD.
_TCHAR *str_subst(_TCHAR *str,													// 09/22/00 DD.
	_TCHAR *oldSubStr,
	_TCHAR *newSubStr,
	_TCHAR *buf);

int str_spell_compare(_TCHAR *str, _TCHAR *candidateStr);                           // 09/26/00 DD.
void str_spell_candidate(_TCHAR *str, _TCHAR *strList, _TCHAR *buf); // 09/26/00 DD.

long binary_spell(															// 09/27/00 AM.
	_TCHAR *str,				// String to search.
	_TCHAR *arr[],			// Array to search.
	long len,					// Active length of array (excluding 0th elt).
	/*UP*/
	long &
	);
long binary_spell(															// 09/27/00 AM.
	_TCHAR *str,				// String to search.
	_TCHAR *arr[],			// Array to search.
	long ii,					// "Left" end of array.
	long jj,					// "Right" end of array.
	/*UP*/
	long &
	);

_TCHAR *spell_candidates(														// 10/16/00 AM.
	_TCHAR *word,				// The word to be checked.
	_TCHAR *buf,				// Buffer for placing candidates.
	long maxbuf				// Maximum length of buffer.
	);
_TCHAR *spell_candidates(														// 09/27/00 AM.
	_TCHAR *word,				// The word to be checked.
	_TCHAR *arr[],			// Alpha-sorted, null terminated array of strings.
	long len,				// Length of array (excluding null terminator string).
	_TCHAR *buf,				// Buffer for placing candidates.
	long maxbuf,
	/*UP*/
	_TCHAR* &ptr				// End of active buffer.					// 10/16/00 AM.
	);

long prefix_match_len(_TCHAR *str1, _TCHAR *str2);						// 09/27/00 AM.
long set_arr_len(_TCHAR *arr[]);											// 09/27/00 AM.

_TCHAR *leftjustify(															// 10/02/00 AM.
	_TCHAR *str,
	long fieldsize
	);
_TCHAR *rightjustify(															// 10/01/00 AM.
	long num,
	long fieldsize
	);

inline bool spell_new_low(													// 10/16/00 AM.
	int weight,				// Current candidate weight
	long len,				// Current candidate length
	int low_weight,		// Best so far
	long low_len,			// Best so far
	long best_len			// Best possible length.
	);
bool str_trim(																	// 11/10/00 DD.
        _TCHAR *str,                      // String to clean.
        /*UP*/
        _TCHAR *buf                       // Supplied buffer to fill.
        );
bool str_ends_with(_TCHAR *str, _TCHAR *ending);							// 11/20/00 DD.

bool str_unescape(_TCHAR*,_TCHAR*,_TCHAR,_TCHAR*);							// 07/26/01 DD.
bool str_escape(_TCHAR*,_TCHAR*,_TCHAR,_TCHAR*);								// 07/26/01 DD.
bool convert_to_float(_TCHAR *str1, _TCHAR *str2, /*UP*/ float &fnum);
bool s_to_f(_TCHAR *str, /*UP*/ float *fnum);							// 08/23/01 AM.

bool str_piece(_TCHAR*,long,long,long,/*UP*/_TCHAR*);					// 11/18/02 AM.

// Should go in a morph module.
bool strsuffix(_TCHAR*,_TCHAR*);												// 11/23/01 AM.
_TCHAR *strstem(_TCHAR*);														// 12/28/01 AM.

bool url_base(_TCHAR*,/*UP*/_TCHAR*);										// 03/16/03 AM.
bool resolve_url(_TCHAR*,_TCHAR*,Parse*,/*DU*/_TCHAR*);					// 05/12/03 AM.
bool unpack_url(																// 05/13/03 AM.
	_TCHAR *url_str,
	Parse *parse,
	/*DU*/
	_TCHAR *scheme,
	_TCHAR *net_loc,
	bool &absflag,
	long &pathlen,
	Dlist<Iarg> *&path,
	bool &endslash,
	_TCHAR *params,
	_TCHAR *query,
	_TCHAR *fragment
	);
bool combine_url(																// 05/14/03 AM.
	_TCHAR *buf,
	_TCHAR *scheme,
	_TCHAR *net_loc,
	bool absflag,
	long pathlen,
	Dlist<Iarg> *path,
	bool endslash,
	_TCHAR *params,
	_TCHAR *query,
	_TCHAR *fragment
	);

long levenshtein_int(			// 03/06/07 AM.
	_TCHAR *s1,
	_TCHAR *s2
	);

#endif
